eoe/man/man1/xargs.1
'\"macro stdmacro
.if n .pH g1.xargs @(#)xargs	30.3 of 3/10/86
.nr X
.if \nX=0 .ds x} XARGS 1 "User Environment Utilities" "\&"
.if \nX=1 .ds x} XARGS 1 "User Environment Utilities"
.if \nX=2 .ds x} XARGS 1 "" "\&"
.if \nX=3 .ds x} XARGS "" "" "\&"
.TH \*(x}
.SH NAME
xargs \- construct argument list(s) and execute command 
.SH SYNOPSIS
.B xargs
[\|options\|] [ command [\|initial-arguments\|] ]
.SH DESCRIPTION
\f4xargs\f1 constructs a command line consisting of the
\f2command\f1
and \f2initial-arguments\f1 
operands specified followed by as many
arguments read in sequence from standard input as will fit 
in length and number constraints specified by the options.
\f4xargs\f1 then invokes the constructed command line and 
waits for its completion. This sequence is repeated until
an end-of-file condition is detected on standard input or
an invocation of a constructed command line returns an
exit status of 255.
.PP
.IR command ,
which may be a shell file,
is searched for, using one's \f3\s-1$PATH\s+1\fP.
If
.I command\^
is omitted,
.B /bin/echo
is used.
.PP
Arguments read in from standard input are defined to be contiguous
strings of characters delimited by one or more blanks, tabs, or new-lines;
empty lines are always discarded.
Blanks and tabs may be embedded as part of an argument if escaped or quoted.
Characters enclosed in quotes (single or double) are taken literally,
and the delimiting quotes are removed.
Outside of quoted strings a backslash
.RB "(" \e ")"
will escape the next character.
.PP
The generated command line length will be the sum of the
size in bytes of the command name and each argument treated 
as strings, including a null byte terminator for each of
these strings. The 
\f4xargs\f1
command will limit the command line length such that when the
command line is invoked, the combined argument and environment
lists will not exceed ARG_MAX-2048 bytes. Within this constraint,
if neither the \f4\-n\f1 nor the \f4\-s\f1 option is specified,
the default command line length will be at least LINE_MAX.
.SH OPTIONS

.TP 10
.BI \-L number\^
.I command\^
is executed for each non-empty
.I number\^
lines of arguments from standard input.
The last invocation of
.I command\^
will be with fewer lines of arguments if fewer than
.I number\^
remain.
A line is considered to end with the first new-line
unless the last character of the line is a blank or a tab;
a trailing blank/tab signals continuation through the next non-empty line.
.TP 10
\f4\-l\f1[\f2number\f1]
Same as -L, except: if
.I number\^
is omitted, 1 is assumed;
option
.B \-x
is forced.
.TP 10
.BI \-I replstr\^
Insert mode:
.I command\^
is executed for each line from standard input,
taking the entire line as a single arg, inserting it in
.I initial-arguments\^
for each occurrence of
.IR replstr .
A maximum of 5 arguments in
.I initial-arguments\^
may each contain one or more instances of
.IR replstr .
Blanks and tabs at the beginning of each line are ignored.
Constructed arguments may not grow larger than what the kernel
.I ncargs
value allows (the default as shipped is 20480), including room
for pointers.  This may be further reduced,
depending on environment size and other arguments),
and option
.B \-x
is forced.
.TP 10
\f4\-i\f1[\f2replstr\f1]\^
Same as \f4\-I\f1 except that
.B "{\|}"
is assumed for
.I replstr\^
if omitted. The \f4\-I\f1 and \f4\-i\f1 options are mutually
exclusive; the last one specified takes effect.
.TP 10
.BI \-n number\^
Execute
.I command\^
using as many standard input arguments as possible, up to
.I number\^
arguments maximum.
Fewer arguments will be used if their total size is greater than
.I size\^
characters, and for the last invocation if there are fewer than
.I number\^
arguments remaining.  In any case, a maximum of 1024 arguments will be
passed per command invocation.
This is further reduced by any extra command
arguments given on the xargs command line.
If option
.B \-x
is also coded, each
.I number\^
arguments must fit in the
.I size\^
limitation, else
.I xargs\^
terminates execution. The \f4\-L\f1, \f4\-l\f1 and \f4\-n\f1
options are mutually exclusive; the last one specified takes
effect.
.TP 10
.B \-t
Trace mode: The
.I command\^
and each constructed argument list are echoed to file descriptor 2
just prior to their execution.
.TP 10
.B \-p
Prompt mode: The user is asked whether to execute
.I command\^
each invocation. Trace mode (\c
.B \-t\c
) is turned on to print the command instance to be executed,
followed by a \f3?.\|.\|.\fP prompt.
A reply of
.B y
(optionally followed by anything) will execute the command; anything else, including
just a carriage return, skips that particular invocation of
.IR command .
.TP 10
.B \-x
Causes
.B xargs
to terminate if any argument list would be greater than
.I size\^
characters;
.B \-x
is forced by the options
.B \-i
and
.BR \-l .
When neither of the options
.BR \-i ,
.BR \-l ,
or
.B \-n
are coded, the total length of all arguments must be within the
.I size\^
limit.
.TP 10
.BI \-s size\^
Invoke command using as many standard input arguments as possible 
yielding a command line length less than \f2size\f1 (a positive
decimal integer) bytes. Fewer arguments will be used if the total
number of arguments (lines) exceeds that specified by the 
\f4\-n\f1 (\f4\-L\f1) option or end of file is encoutered on
standard input before \f2size\f1 bytes are accumulated.
The maximum
.I size\^
depends partly on the size of the environment, but is limited to
a maximum of the kernel value of
.I ncargs
(the default as shipped is 20480), including room
for pointers. If
.B \-s
is not coded, the maximum is taken as the default.
.TP 10
\f4\-E\f2eofstr\f1
Specify a logical end-of-file string to replace the default underscore. 
.BI xargs\^
reads standard input until either end-of-file or the logical 
\f3\s-1EOF\s+1\fP string is encountered.
.TP 10
\f4\-e\f1[\f2eofstr\f1]
Same as \f4\-E\f1 except that
the option
\f3\-e\fP with no
.I eofstr\^
coded turns off the logical \f3\s-1EOF\s+1\fP string capability
(underbar is taken literally).
.PP
Note: The \f4\-e\f1, \f4\-i\f1 and \f4\-l\f1 take optional option-arguments
that cannot be separate arguments.
.PP
.B xargs
will terminate if either it receives a return code of
.B \-1
from,
or if it cannot execute,
.IR command .
When
.I command\^
is a shell program,
it should explicitly
.I exit\^
(see
.IR sh (1))
with an appropriate value to avoid accidentally returning with
.BR \-1 .
.SH EXAMPLES
The following will move all files from directory $1 to directory $2, and echo each move
command just before doing it:
.PP
.RS 10
ls \|$1 \|| \|xargs \|\-i \|\-t \|mv \|$1/{\|} \|$2/{\|}
.RE
.PP
or
.RS 10
ls \|$1 \|| \|xargs \|\-I /{\|} \|\-t \|mv \|$1/{\|} \|$2/{\|}
.RE
.PP
The following will combine the output of the parenthesized commands onto one line,
which is then echoed to the end of file \f2log\fP:
.PP
.RS 10
(logname; \|date; \|echo \|$0 \|$\(**) \|| \|xargs \|>>log
.RE
.PP
The user is asked which files in the current directory are to be archived
and archives them into \f2arch\fP (1.) one at a time, or (2.) many at a time.
.PP
.RS 10
1.\ \|\ \|ls \|| \|xargs \|\-p \|\-l \|ar \|r \|arch
.br
2.\ \|\ \|ls \|| \|xargs \|\-p \|\-l \|| \|xargs \|ar \|r \|arch
.RE
.PP
The following will execute
.IR diff "(1)"
with successive
pairs of arguments originally typed as shell arguments:
.PP
.RS 10
echo \|$\(** \|| \|xargs \|\-n2 \|diff
.RE
.SH DIAGNOSTICS
\f4xargs\f1
exits with 0 when all invocations of command returned exit status zero.
\f4xargs\f1 exits with 126 when the command specified was found but 
could not be invoked. \f4xargs\f1 exits with 127 when the command specified
could not be found.
.SH "SEE ALSO"
sh(1).
.Ee


eoe/cmd/xargs/xargs.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#include <sys/wait.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

#ident	"cmd/xargs/xargs.c: $Revision: 1.14 $"
#define FALSE 0
#define TRUE 1

/*
 * want as large as possible, but sysconf(_SC_ARG_MAX) includes
 * environment and word for a pointer per arg also.
 */
#define MAXARGS 1024
#define MAXINSERTS 5
char Errstr[256];
char *arglist[MAXARGS+1];
char *argbuf;
char *next;
char *buffer;
char *lastarg = "";
char **ARGV = arglist;
char *LEOF = "_"; 
char *INSPAT = "{}";
struct inserts {
	char **p_ARGV;		/* where to put newarg ptr in arg list */
	char *p_skel;		/* ptr to arg template */
	} saveargv[MAXINSERTS];
char *ins_buf;
char *p_ibuf;
int PROMPT = -1;
int BUFLIM;	/* set at top of main now. */
int Maxstdinargs;	/* max args that can be picked up from stdin.
	set to MAXARGS - cmdargs from xargs cmd line */
int N_ARGS = 0;
int N_args = 0;
int N_lines = 0;
int DASHX = FALSE;
int MORE = TRUE;
int PER_LINE = FALSE;
int ERR = FALSE;
int OK = TRUE;
int LEGAL = FALSE;
int TRACE = FALSE;
int INSERT = FALSE;
int linesize = 0;
int ibufsize = 0;
int Ncargs;
int Maxibuf;
int Maxsbuf;
void ermsg(char *);
void addibuf(struct inserts *);

main(argc,argv)
int argc;
char **argv; 
{
    extern char *addarg(), *getarg(), *checklen(), *insert();
    char *cmdname, *initbuf, **initlist, *flagval;
    int  initsize;
    int  retval = 0;
    register int j, n_inserts;
    register struct inserts *psave;
    
    /* initialization */

    argc--; argv++;
    n_inserts = 0;
    psave = saveargv;

    j = getenvsize();	/* need to do this before arg processing */
    Ncargs = sysconf(_SC_ARG_MAX);
    Maxibuf = Ncargs-1024;
    Maxsbuf = Ncargs/2;

    if (((argbuf  = (char *)malloc(Ncargs)) == NULL) ||
	((ins_buf = (char *)malloc(Maxibuf)) == NULL) ||
	((buffer  = (char *)malloc(Maxsbuf)) == NULL)) {
	sprintf(Errstr, "Cannot allocate more memory\n");
	ermsg(Errstr);
	exit(1);
    }
    next = argbuf;

    BUFLIM = Ncargs - j;

    if(BUFLIM <= ((int)sizeof(char *)+32)) {
	sprintf(Errstr, "Environment is too large (%d); no room left for arguments\n",
		j);
	ermsg(Errstr);
	exit(1);
    }

    /* look for flag arguments */

    while  ( *argv && (*argv)[0] == '-'  ) {
	if(!strcmp(*argv, "--")) {
	  argv++;
	  argc--;
	  break;
	}
	flagval = *argv+1;
	switch ( *flagval++ ) {
	  case 'x': DASHX = LEGAL = TRUE;
	    break;

	  case 'l': 
	    PER_LINE = LEGAL = TRUE;
	    N_ARGS = 0;
	    INSERT = FALSE;
	    if( *flagval && (PER_LINE=atoi(flagval)) <= 0 ) {
		sprintf(Errstr, "#lines must be positive int: %s\n", *argv);
		ermsg(Errstr);
	    }
	    break;

	  case 'L':
	    PER_LINE = TRUE;
	    N_ARGS = 0;
	    INSERT = FALSE;
	    if(!(*flagval)){
		flagval = *(++argv);
		argc--;
	    }
	    if((PER_LINE=atoi(flagval)) <= 0 ) {
		sprintf(Errstr, "#lines must be positive int: %s\n", *argv);
		ermsg(Errstr);
	    }
	    break;
	    

	  case 'i': 
	    INSERT = PER_LINE = LEGAL = TRUE;
	    N_ARGS = 0;
	    if ( *flagval ) 
		INSPAT = flagval;
	    break;

	  case 'I':
	    INSERT = PER_LINE = LEGAL = TRUE;
	    N_ARGS = 0;
	    if( *flagval)
	      INSPAT = flagval;
	    else{
		INSPAT = *(++argv);
		argc--;
	    }
	    break;

	  case 't': TRACE = TRUE;
	    break;

	  case 'e': LEOF = flagval;
	    break;

	  case 'E':
	    if(*flagval)
	      LEOF = flagval;
	    else{
		LEOF = *(++argv);
		argc--;
	    }
	    break;

	  case 's': 
	    if(!(*flagval)){
		flagval = *++argv;
		argc--;
	    }
	    j = atoi(flagval);
	    if( j>=BUFLIM  ||  j<=0 ) {
		sprintf(Errstr, "%s invalid: must be > 0 and < %d (varies with environment)\n", *argv, BUFLIM);
		ermsg(Errstr);
	    }
	    else
	      BUFLIM = j;
	    break;

	  case 'n': 
	    if(!(*flagval)){
		flagval = *++argv;
		argc--;
	    }
	    if( (N_ARGS = atoi(flagval)) <= 0 ) {
		sprintf(Errstr, "#args must be positive int: %s\n", *argv);
		ermsg(Errstr);
	    }
	    else {
		LEGAL = DASHX || N_ARGS==1;
		INSERT = PER_LINE = FALSE;
	    }
	    break;

	  case 'p': 
	    if( (PROMPT = open("/dev/tty",0)) == -1) {
		ermsg("can't read from tty for -p\n");
	    }
	    else
	      TRACE = TRUE;
	    break;

	  default: 
	    sprintf(Errstr, "unknown option: %s\n", *argv);
	    ermsg(Errstr);
	    break;
	}

	argv++;
	if ( --argc < 1 ) break;
    }

    if( ! OK )
      ERR = TRUE;

    /* pick up command name */
    
    if ( argc == 0 ) {
	cmdname = "/bin/echo";
	*ARGV++ = addarg(cmdname);
    }
    else
      cmdname = *argv;
    
    /* pick up args on command line */

    while ( OK && argc-- ) {
	if ( INSERT && ! ERR ) {
	    if ( xindex(*argv, INSPAT) != -1 ) {
		if ( ++n_inserts > MAXINSERTS ) {
		    sprintf(Errstr, "too many args with %s\n", INSPAT);
		    ermsg(Errstr);
		    ERR = TRUE;
		}
		psave->p_ARGV = ARGV;
		(psave++)->p_skel = *argv;
	    }
	}
	*ARGV++ = addarg( *argv++ );
	if(ARGV == &arglist[MAXARGS-10])  {	/* leave some room for stdin */
	    sprintf(Errstr, "too many xargs command line arguments (%d max)\n",
		    MAXARGS-10);
	    ermsg(Errstr);
	    exit(1);
	}
    }
    
    Maxstdinargs = &arglist[MAXARGS-1] - ARGV;

    /* pick up args from standard input */

    initbuf = next;
    initlist = ARGV;
    initsize = linesize;
    
    while ( OK && MORE ) {
	next = initbuf;
	ARGV = initlist;
	linesize = initsize;
	if ( *lastarg )
	  *ARGV++ = addarg( lastarg );

	while ( (*ARGV++ = getarg()) && OK );

	/* insert arg if requested */

	if ( !ERR && INSERT ) {
	    p_ibuf = ins_buf;
	    ARGV--;
	    j = ibufsize = 0;
	    for ( psave=saveargv;  ++j<=n_inserts;  ++psave ) {
		addibuf(psave);
		if ( ERR ) break;
	    }
	}
	*ARGV = 0;

				/* exec command */

	if ( ! ERR ) {
	    if ( ! MORE && (PER_LINE && N_lines==0 || N_ARGS && N_args==0) ) exit (0);
	    OK = TRUE;
	    j = TRACE ? echoargs() : TRUE;
	    if( j ) {
		int rc;
		rc = lcall(cmdname, arglist);
		retval |= rc;
		if ( rc != -1 ) continue;
		sprintf(Errstr, "%s not executed, returned -1 or killed\n", cmdname);
		ermsg(Errstr);
	    }
	}
    }
    if ( OK && !retval) return (0); else return (1);
}

char *
checklen(arg)
char *arg;
{
    register int oklen;

    oklen = TRUE;
    if ( (linesize += strlen(arg)+1) > BUFLIM ) {
	lastarg = arg;
	oklen = OK = FALSE;
	if ( LEGAL ) {
	    ERR = TRUE;
	    ermsg("arg list too long\n");
	}
	else if( N_args > 1 )
	  N_args = 1;
	else {
	    sprintf(Errstr, "single arg or initial list was > than max length (%d) with environment\n",
		    BUFLIM);
	    ermsg(Errstr);
	    ERR = TRUE;
	}
    }
    return ( oklen  ? arg : 0 );
}

char *
addarg(arg)
char *arg;
{
	strcpy(next, arg);
	arg = next;
	next += strlen(arg)+1;
	return ( checklen(arg) );
}

char *
getarg()
{
    register char c, c1, *arg;
    char *retarg;
    
    while ( (c=getchr()) == ' '
	   || c == '\n'
	   || c == '\t' );
    if ( c == '\0' ) {
	MORE = FALSE;
	return 0;
    }

    arg = next;
    for ( ; ; c = getchr() )
      switch ( c ) {
	  
	case '\t':
	case ' ' :
	  if ( INSERT ) { *next++ = c;
			  break;
		      }
	  
	case '\0':
	case '\n':
	  *next++ = '\0';
	  if( !strcmp(arg,LEOF) || c=='\0' ) {
	      MORE = FALSE;
	      if( c != '\n' )
		while( c=getchr() )
		  if( c=='\n' ) break;
	      return 0;
	  }
	  else {
	      ++N_args;
	      if( retarg = checklen(arg) ) {
		  if( (PER_LINE && c=='\n' && ++N_lines>=PER_LINE)
		     ||   (N_ARGS && N_args>=N_ARGS) ||
		     N_args >= Maxstdinargs) {
		      N_lines = N_args = 0;
		      lastarg = "";
		      OK = FALSE;
		  }
	      }
	      return retarg;
	  }
	  
	case '\\':
	  *next++ = getchr();
	  break;

	case '"':
	case '\'':
	  while( (c1=getchr()) != c) {
	      if( c1 == '\0' || c1 == '\n' ) {
		  *next++ = '\0';
		  sprintf(Errstr, "missing quote?: %s\n", arg);
		  ermsg(Errstr);
		  ERR = TRUE;
		  return (0);
	      }
	      *next++ = c1;
	  }
	  break;
	  
	default:
	  *next++ = c;
	  break;
      }
}

void
ermsg(messages)
char *messages;
{
    write(2,"xargs: ",7);
    write(2,messages,strlen(messages));
    OK = FALSE;
}

echoargs()
{
    register char **anarg;
    char yesorno[1], junk[1];
    register int j;

    anarg = arglist;
    anarg--;
    while ( *++anarg ) {
	write(2, *anarg, strlen(*anarg) );
	write(2," ",1);
    }
    if( PROMPT == -1 ) {
	write(2,"\n",1);
	return TRUE;
    }
    write(2,"?...",4);
    if( read(PROMPT,yesorno,1) == 0 )
      exit(0);
    if( yesorno[0] == '\n' )
	return FALSE;
    while( ((j=read(PROMPT,junk,1))==1) && (junk[0]!='\n') );
    if( j==0 )
	exit (0);
    return ( yesorno[0]=='y' );
}

char *
insert(pattern, subst)
char *pattern, *subst;
{
    int len, ipatlen;
    register char *pat;
    register char *bufend;
    register char *pbuf;

    len = strlen(subst);
    ipatlen = strlen(INSPAT)-1;
    pat = pattern-1;
    pbuf = buffer;
    bufend = &buffer[Maxsbuf];
    
    while ( *++pat ) {
	if( xindex(pat,INSPAT) == 0 ) {
	    if ( pbuf+len >= bufend ) break;
	    else {
		strcpy(pbuf, subst);
		pat += ipatlen;
		pbuf += len;
	    }
	}
	else {
	    *pbuf++ = *pat;
	    if (pbuf >= bufend ) break;
	}
    }
    
    if ( ! *pat ) {
	*pbuf = '\0';
	return (buffer);
    }
    else {
	sprintf(Errstr, "max arg size with insertion via %s's exceeded\n", INSPAT);
	ermsg(Errstr);
	ERR = TRUE;
	return 0;
    }
}

void
addibuf(p)
struct inserts *p;
{
    register char *newarg, *skel, *sub;
    int l;

    skel = p->p_skel;
    sub = *ARGV;
    linesize -= strlen(skel)+1;
    newarg = insert(skel,sub ? sub : "");
    if ( checklen(newarg) ) {
	if( (ibufsize += (l=strlen(newarg)+1)) > Maxibuf) {
	    ermsg("insert-buffer overflow\n");
	    ERR = TRUE;
	}
	strcpy(p_ibuf, newarg);
	*(p->p_ARGV) = p_ibuf;
	p_ibuf += l;
    }
}

getchr() {
    int c;
    c = getc(stdin);
    if (c == EOF)
	return(0);
    else
	return(c);
}

int lcall(sub,subargs)
char *sub, **subargs;
{

	int retcode;
	register int iwait, child;

	switch( child=fork() ) {
	default:
		while((iwait=waitpid(child, &retcode, 0)) == -1 && errno == ECHILD) ;
		if (iwait == -1
		 || WIFSIGNALED(retcode) && WTERMSIG(retcode) != SIGPIPE
		 || WIFEXITED(retcode)	&& WEXITSTATUS(retcode) == 0xff
		)
			return -1;
		if(WIFEXITED(retcode)){
		    if (WEXITSTATUS(retcode) == 126)
		      exit(126);
		    else if (WEXITSTATUS(retcode) == 127)
		      exit(127);
		}

		return WEXITSTATUS(retcode);
	case 0:
		execvp(sub,subargs);
		if(errno == EACCES) exit(126);
		if(errno == ENOENT) exit(127);
		return (0);
	case -1:
		return (-1);
		}
}
/*
	If `s2' is a substring of `s1' return the offset of the first
	occurrence of `s2' in `s1',
	else return -1.
*/

xindex(as1,as2)
char *as1,*as2;
{
	register char *s1,*s2,c;
	int offset;

	s1 = as1;
	s2 = as2;
	c = *s2;

	while (*s1)
		if (*s1++ == c) {
			offset = s1 - as1 - 1;
			s2++;
			while ((c = *s2++) == *s1++ && c) ;
			if (c == 0)
				return(offset);
			s1 = offset + as1 + 1;
			s2 = as2;
			c = *s2;
		}
	 return(-1);
}

/* figure out size of environment, including ptrs, so we
 * won't hit the sysconf(_SC_ARG_MAX) limit, but can still do as many
 * args as possible per exec.
*/
getenvsize()
{
	extern char **environ;
	char **e = environ;
	int len = 0;

	while(*e) {
		len += sizeof(*e);
		len += strlen(*e);
		e++;
	}
	return len;
}
